### PR TITLE
Nucleo f031

### PR DESCRIPTION
**Pull request by Gimpel-K67**
This branch adds support for the nucleo_f031k6 board.
The configuration is based on the f303k8 and the ST reference manual.

I had successfully tested the following sample code:
blinky
blink_led (uses TIM2_CH2 on PB3)
fade_led
hello_world

### CODE REVIEW

File: boards/arm/nucleo_f031k6/doc/index.rst [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+Hardware
+********
+Nucleo F031K6 provides the following hardware components:
+
+- STM32 microcontroller in LQFP32 package
+- One types of extension resources:
+
+  - Arduino* nano connectivity
+
+- ARM* mbed*
```
Comment (erwango) [inline boards/arm/nucleo_f031k6/doc/index.rst:36]: Not hardware related.

File: boards/arm/nucleo_f031k6/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+  - USB communication (LD1), user LED (LD2), power LED (LD3)
+
+- reset push button
+
+- Support of wide choice of Integrated Development Environments (IDEs) including:
+
+  - IAR
+  - ARM Keil
+  - GCC-based IDEs
```
Comment (erwango) [inline boards/arm/nucleo_f031k6/doc/index.rst:53]: Not hardware related

File: boards/arm/nucleo_f031k6/nucleo_f031k6_defconfig [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# Zephyr Kernel Configuration
+CONFIG_SOC_SERIES_STM32F0X=y
```
Comment (erwango) [inline boards/arm/nucleo_f031k6/nucleo_f031k6_defconfig:4]: Nit: I'd put this as part of platform configuration.

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this contribution. Few minor comments, otherwise LGTM.
Pretty small target but interesting challenge!

File: soc/arm/st_stm32/stm32f0/Kconfig.defconfig.stm32f031x6 [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+# Copyright (c) 2021 Sebastian Schwabe <sebastian.schwabe@mailbox.tu-dresden.de>
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_STM32F031X6
+
+config SOC
+	default "stm32f031x6"
+
+config NUM_IRQS
+	default 28
```
Comment (FRASTM) [inline soc/arm/st_stm32/stm32f0/Kconfig.defconfig.stm32f031x6:12]: could you confirm 28 (RM0091 mentioning 32 maskable interrupt channels)

**Review by FRASTM — APPROVED** [review]

File: soc/arm/st_stm32/stm32f0/Kconfig.defconfig.stm32f031x6 (hunk 4 above)
Comment (TeckKitty) [reply soc/arm/st_stm32/stm32f0/Kconfig.defconfig.stm32f031x6:12]: I referred to issue #16716 and [this comment](https://github.com/zephyrproject-rtos/zephyr/issues/16716#issuecomment-500679521).  In `modules/hal/stm32/stm32cube/stm32f0xx/soc/stm32f031x6.h` the highest number is 27 so I set 28 for NUM_IRQ. I did not find an exact guide for STM32 chips in the doc so I thought the information in the issue is fine.

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
To be synched with 34362

**Comment by MaureenHelm** [comment]
> To be synched with 34362

Is it ok to merge this PR, or does 34362 need to go in first?

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**